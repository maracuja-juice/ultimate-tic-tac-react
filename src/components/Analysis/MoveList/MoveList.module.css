.noPadding {
  padding: 0 !important;
}

.fullWidth {
  width: 100%;
}

.smallerMargin :global(.mdc-list-item__graphic) {
  margin-right: 15px !important;
}
