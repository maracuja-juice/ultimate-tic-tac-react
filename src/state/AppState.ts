import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
