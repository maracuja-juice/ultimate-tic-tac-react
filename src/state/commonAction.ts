export const RESTART_GAME = 'common/restart-game';

export const restartGame = () => ({
    type: RESTART_GAME
});