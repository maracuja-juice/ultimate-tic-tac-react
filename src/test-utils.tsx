// See: https://testing-library.com/docs/react-testing-library/setup#custom-render
import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IconProvider } from './components/Icons';
import '@testing-library/jest-dom'; // https://github.com/testing-library/jest-dom/issues/282#issuecomment-660034778
import { MemoryRouter } from 'react-router-dom';

const AllTheProviders: FC = ({ children }) => {
  return (
    <IconProvider>
      <MemoryRouter>{children}</MemoryRouter>
    </IconProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
