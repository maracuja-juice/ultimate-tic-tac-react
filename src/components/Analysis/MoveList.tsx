import * as React from 'react';
import { List, SimpleListItem } from '@rmwc/list';
import { CustomEventT } from '@rmwc/types';
import { Move, Player } from '../../state/AppState';
import { ReactNode } from 'react';

interface MoveListProps {
  reversedMoves: Move[];
  currentMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
}

export class MoveList extends React.Component<MoveListProps> {
  playerAsString = (player: Player) => {
    if (player === Player.Cross) {
      return 'x';
    } else if (player === Player.Circle) {
      return 'o';
    }
    return undefined;
  };

  getMoves = () => {
    const { reversedMoves, currentMove } = this.props;
    const moveList: ReactNode[] = [];

    reversedMoves.forEach((m: Move) => {
      moveList.push(
        <SimpleListItem
          key={m.moveNumber}
          activated={currentMove === m.moveNumber}
          graphic={{ icon: this.playerAsString(m.player), size: 'medium' }}
          text={'Move ' + m.moveNumber}
          secondaryText={
            'Board ' +
            m.boardPosition.x +
            '/' +
            m.boardPosition.y +
            ' - ' +
            'Field ' +
            m.tilePosition.x +
            '/' +
            m.tilePosition.y
          }
        />,
      );
    });

    return moveList;
  };

  changeDisplayedMove = (event: CustomEventT<number>) => {
    const {
      currentMove,
      reversedMoves,
      moveForwardInHistory,
      moveBackwardInHistory,
    } = this.props;

    const numberOfMovesFromEnd = event.detail;
    const moveNumber = reversedMoves.length - numberOfMovesFromEnd;

    if (moveNumber === currentMove) {
      return;
    }

    const amountOfMovesToMove = moveNumber - currentMove;
    if (amountOfMovesToMove <= -1) {
      moveBackwardInHistory(currentMove - moveNumber);
    } else {
      moveForwardInHistory(moveNumber - currentMove);
    }
  };

  render() {
    return (
      <List twoLine={true} dense={true} onAction={this.changeDisplayedMove}>
        {this.getMoves()}
      </List>
    );
  }
}
