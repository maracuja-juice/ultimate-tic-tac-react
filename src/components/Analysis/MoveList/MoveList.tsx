import * as React from 'react';
import { List, SimpleListItem } from '@rmwc/list';
import { CustomEventT } from '@rmwc/types';
import { MoveState, Player } from '../../../state/AppState';
import { Element } from 'react-scroll';
import { moveScrollElementBaseName } from '../ScrollElementConstants';
import styles from './MoveList.module.css';
import classNames from 'classnames';
import { OIcon, XIcon } from '../../Icons';

export interface MoveListProps {
  reversedMoves: MoveState[];
  currentMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
}

export function MoveList(props: MoveListProps) {
  const {
    currentMove,
    reversedMoves,
    moveForwardInHistory,
    moveBackwardInHistory,
  } = props;

  const changeDisplayedMove = (event: CustomEventT<{ index: number }>) => {
    const numberOfMovesFromEnd = event.detail.index;
    const moveNumber = reversedMoves.length - numberOfMovesFromEnd;

    if (moveNumber === currentMove) {
      return;
    }

    const amountOfMovesToMove = moveNumber - currentMove;
    if (amountOfMovesToMove < 0) {
      moveBackwardInHistory(currentMove - moveNumber);
    } else {
      moveForwardInHistory(moveNumber - currentMove);
    }
  };

  return (
    <List
      twoLine={true}
      dense={true}
      onAction={changeDisplayedMove}
      className={styles.noPadding}
    >
      {reversedMoves.map((m) => {
        return (
          <Element
            key={m.moveNumber}
            name={moveScrollElementBaseName + m.moveNumber}
          >
            <SimpleListItem
              tag="button"
              activated={currentMove === m.moveNumber}
              graphic={{
                icon: m.player === Player.Cross ? <XIcon /> : <OIcon />,
                size: 'medium',
                'aria-label': `Player ${
                  m.player === Player.Cross ? 'X' : 'O'
                } Icon`,
              }}
              text={'Move ' + m.moveNumber}
              secondaryText={`Board ${m.boardPosition.x}/${m.boardPosition.y} - Tile ${m.tilePosition.x}/${m.tilePosition.y}`}
              className={classNames([
                styles.smallerMargin,
                styles.biggerIcon,
                styles.fullWidth,
              ])}
            />
          </Element>
        );
      })}
    </List>
  );
}
