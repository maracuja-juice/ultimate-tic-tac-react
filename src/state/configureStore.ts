import { AppState } from './AppState';
import { applyMiddleware, combineReducers, createStore } from 'redux';
import logger from 'redux-logger';
import tileReducer from './tiles/tileReducer';
import { composeWithDevTools } from 'redux-devtools-extension';
import gameReducer from './game/gameReducer';
import moveReducer from './moves/moveReducer';
import { enableBatching } from 'redux-batched-actions';
import activeBoardReducer from './activeBoards/activeBoardsReducer';
import createSagaMiddleware from 'redux-saga';
import { all, fork } from 'redux-saga/effects';
import moveValidationSaga from './sagas/MoveValidationSaga';

const rootreducer = combineReducers<AppState>(
    {
        game: gameReducer,
        tiles: tileReducer,
        moves: moveReducer,
        activeBoards: activeBoardReducer
    } );

export function configureStore() {

    const sagaMiddleware = createSagaMiddleware();

    const middleWaresToApply = [
        logger,
        sagaMiddleware
    ];
    const middleware = applyMiddleware( ...middleWaresToApply );
    const store = createStore( enableBatching( rootreducer ), composeWithDevTools(
        middleware
    ) );

    sagaMiddleware.run( rootSaga );

    return store;
}

function* rootSaga() {
    yield all(
        [
            fork( moveValidationSaga ),
        ]
    );
}