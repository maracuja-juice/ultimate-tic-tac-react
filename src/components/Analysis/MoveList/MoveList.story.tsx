import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { action } from '@storybook/addon-actions';
import { MoveList, MoveListProps } from './MoveList';
import { movesForBoardWithThreeMovesMock } from '../../../__mocks__';
import { IconProvider } from '../../IconProvider';

export default {
  title: 'MoveList',
  component: MoveList,
} as Meta;

export const Default: Story<MoveListProps> = (args) => (
  <IconProvider>
    <div className="moveList">
      <MoveList
        {...args}
        moveForwardInHistory={action('moveForwardInHistory')}
        moveBackwardInHistory={action('moveBackwardInHistory')}
      />
    </div>
  </IconProvider>
);

Default.args = {
  currentMove: 2,
  reversedMoves: movesForBoardWithThreeMovesMock.slice().reverse(),
};
