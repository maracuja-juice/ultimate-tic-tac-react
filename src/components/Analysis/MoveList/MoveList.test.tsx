import * as React from 'react';
import { render, screen } from '../../../test-utils';
import userEvent from '@testing-library/user-event';
import { MoveState, Player } from '../../../state/AppState';
import { MoveList } from './MoveList';

const moveStates: MoveState[] = [
  {
    moveNumber: 1,
    player: Player.Cross,
    tilePosition: { x: 0, y: 0 },
    boardPosition: { x: 0, y: 0 },
  },
  {
    moveNumber: 2,
    player: Player.Circle,
    tilePosition: { x: 0, y: 1 },
    boardPosition: { x: 0, y: 0 },
  },
  {
    moveNumber: 3,
    player: Player.Cross,
    tilePosition: { x: 0, y: 0 },
    boardPosition: { x: 0, y: 1 },
  },
].reverse();

test('renders moves', () => {
  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={1}
      onMoveUpwardsInList={() => {}}
      onMoveDownwardsInList={() => {}}
    />,
  );

  moveStates.map((value) => {
    const regex = new RegExp(
      `move ${value.moveNumber} board ` +
        `${value.boardPosition.x}\/${value.boardPosition.y} ` +
        `- tile ${value.tilePosition.x}\/${value.tilePosition.y}`,
      'i',
    );
    expect(screen.getByRole('button', { name: regex })).toBeInTheDocument();
  });
});

test('renders currentMove as activated', () => {
  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={2}
      onMoveUpwardsInList={() => {}}
      onMoveDownwardsInList={() => {}}
    />,
  );

  const moveTwo = screen.getByRole('button', { name: /move 2/i });
  expect(moveTwo.className).toMatch(/activated/);
});

test('allows using buttons to move upwards in list', () => {
  const moveUpwardsInList = jest.fn();
  const moveDownwardsInList = jest.fn();

  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={3}
      onMoveUpwardsInList={moveUpwardsInList}
      onMoveDownwardsInList={moveDownwardsInList}
    />,
  );

  const moveOne = screen.getByRole('button', { name: /move 1/i });
  const moveTwo = screen.getByRole('button', { name: /move 2/i });

  userEvent.click(moveOne);

  expect(moveUpwardsInList).toHaveBeenCalledWith(2);
  expect(moveDownwardsInList).not.toHaveBeenCalled();

  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={3}
      onMoveUpwardsInList={moveUpwardsInList}
      onMoveDownwardsInList={moveDownwardsInList}
    />,
  );

  userEvent.click(moveTwo);

  expect(moveUpwardsInList).toHaveBeenCalledWith(1);
  expect(moveDownwardsInList).not.toHaveBeenCalled();
});

test('allows using buttons to move downwards in list', () => {
  const moveUpwardsInList = jest.fn();
  const moveDownwardsInList = jest.fn();

  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={1}
      onMoveUpwardsInList={moveUpwardsInList}
      onMoveDownwardsInList={moveDownwardsInList}
    />,
  );

  const moveTwo = screen.getByRole('button', { name: /move 2/i });
  const moveThree = screen.getByRole('button', { name: /move 3/i });

  userEvent.click(moveThree);
  expect(moveDownwardsInList).toHaveBeenCalledWith(2);
  expect(moveUpwardsInList).not.toHaveBeenCalled();

  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={1}
      onMoveUpwardsInList={moveUpwardsInList}
      onMoveDownwardsInList={moveDownwardsInList}
    />,
  );

  userEvent.click(moveTwo);
  expect(moveDownwardsInList).toHaveBeenCalledWith(1);
  expect(moveUpwardsInList).not.toHaveBeenCalled();
});

test('doesnt call callbacks when clicking on active item', () => {
  const moveUpwardsInList = jest.fn();
  const moveDownwardsInList = jest.fn();

  render(
    <MoveList
      reversedMoves={moveStates}
      activatedMove={1}
      onMoveUpwardsInList={moveUpwardsInList}
      onMoveDownwardsInList={moveDownwardsInList}
    />,
  );

  const moveOne = screen.getByRole('button', { name: /move 1/i });
  userEvent.click(moveOne);

  expect(moveDownwardsInList).not.toHaveBeenCalled();
  expect(moveUpwardsInList).not.toHaveBeenCalled();
});
