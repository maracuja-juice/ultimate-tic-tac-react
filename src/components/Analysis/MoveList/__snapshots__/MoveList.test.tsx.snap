// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveList should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MoveList
    currentMove={4}
    moveBackwardInHistory={[Function]}
    moveForwardInHistory={[Function]}
    reversedMoves={
        Array [
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 4,
              "player": 1,
              "tilePosition": Object {
                "x": 0,
                "y": 2,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 1,
              },
              "moveNumber": 3,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 2,
              "player": 1,
              "tilePosition": Object {
                "x": 0,
                "y": 1,
              },
            },
            Object {
              "boardPosition": Object {
                "x": 0,
                "y": 0,
              },
              "moveNumber": 1,
              "player": 0,
              "tilePosition": Object {
                "x": 0,
                "y": 0,
              },
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Element
          name="move-4"
>
          <SimpleListItem
                    activated={true}
                    graphic={
                              Object {
                                        "icon": "o",
                                        "size": "small",
                                      }
                    }
                    secondaryText="Board 0/0 - Field 0/2"
                    text="Move 4"
          />
</Element>,
        <Element
          name="move-3"
>
          <SimpleListItem
                    activated={false}
                    graphic={
                              Object {
                                        "icon": "x",
                                        "size": "small",
                                      }
                    }
                    secondaryText="Board 0/1 - Field 0/0"
                    text="Move 3"
          />
</Element>,
        <Element
          name="move-2"
>
          <SimpleListItem
                    activated={false}
                    graphic={
                              Object {
                                        "icon": "o",
                                        "size": "small",
                                      }
                    }
                    secondaryText="Board 0/0 - Field 0/1"
                    text="Move 2"
          />
</Element>,
        <Element
          name="move-1"
>
          <SimpleListItem
                    activated={false}
                    graphic={
                              Object {
                                        "icon": "x",
                                        "size": "small",
                                      }
                    }
                    secondaryText="Board 0/0 - Field 0/0"
                    text="Move 1"
          />
</Element>,
      ],
      "dense": true,
      "onAction": [Function],
      "twoLine": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "4",
        "nodeType": "class",
        "props": Object {
          "children": <SimpleListItem
            activated={true}
            graphic={
                        Object {
                                    "icon": "o",
                                    "size": "small",
                                  }
            }
            secondaryText="Board 0/0 - Field 0/2"
            text="Move 4"
/>,
          "name": "move-4",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activated": true,
            "graphic": Object {
              "icon": "o",
              "size": "small",
            },
            "secondaryText": "Board 0/0 - Field 0/2",
            "text": "Move 4",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "children": <SimpleListItem
            activated={false}
            graphic={
                        Object {
                                    "icon": "x",
                                    "size": "small",
                                  }
            }
            secondaryText="Board 0/1 - Field 0/0"
            text="Move 3"
/>,
          "name": "move-3",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activated": false,
            "graphic": Object {
              "icon": "x",
              "size": "small",
            },
            "secondaryText": "Board 0/1 - Field 0/0",
            "text": "Move 3",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "children": <SimpleListItem
            activated={false}
            graphic={
                        Object {
                                    "icon": "o",
                                    "size": "small",
                                  }
            }
            secondaryText="Board 0/0 - Field 0/1"
            text="Move 2"
/>,
          "name": "move-2",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activated": false,
            "graphic": Object {
              "icon": "o",
              "size": "small",
            },
            "secondaryText": "Board 0/0 - Field 0/1",
            "text": "Move 2",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "children": <SimpleListItem
            activated={false}
            graphic={
                        Object {
                                    "icon": "x",
                                    "size": "small",
                                  }
            }
            secondaryText="Board 0/0 - Field 0/0"
            text="Move 1"
/>,
          "name": "move-1",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activated": false,
            "graphic": Object {
              "icon": "x",
              "size": "small",
            },
            "secondaryText": "Board 0/0 - Field 0/0",
            "text": "Move 1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Element
            name="move-4"
>
            <SimpleListItem
                        activated={true}
                        graphic={
                                    Object {
                                                "icon": "o",
                                                "size": "small",
                                              }
                        }
                        secondaryText="Board 0/0 - Field 0/2"
                        text="Move 4"
            />
</Element>,
          <Element
            name="move-3"
>
            <SimpleListItem
                        activated={false}
                        graphic={
                                    Object {
                                                "icon": "x",
                                                "size": "small",
                                              }
                        }
                        secondaryText="Board 0/1 - Field 0/0"
                        text="Move 3"
            />
</Element>,
          <Element
            name="move-2"
>
            <SimpleListItem
                        activated={false}
                        graphic={
                                    Object {
                                                "icon": "o",
                                                "size": "small",
                                              }
                        }
                        secondaryText="Board 0/0 - Field 0/1"
                        text="Move 2"
            />
</Element>,
          <Element
            name="move-1"
>
            <SimpleListItem
                        activated={false}
                        graphic={
                                    Object {
                                                "icon": "x",
                                                "size": "small",
                                              }
                        }
                        secondaryText="Board 0/0 - Field 0/0"
                        text="Move 1"
            />
</Element>,
        ],
        "dense": true,
        "onAction": [Function],
        "twoLine": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "4",
          "nodeType": "class",
          "props": Object {
            "children": <SimpleListItem
              activated={true}
              graphic={
                            Object {
                                          "icon": "o",
                                          "size": "small",
                                        }
              }
              secondaryText="Board 0/0 - Field 0/2"
              text="Move 4"
/>,
            "name": "move-4",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activated": true,
              "graphic": Object {
                "icon": "o",
                "size": "small",
              },
              "secondaryText": "Board 0/0 - Field 0/2",
              "text": "Move 4",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "children": <SimpleListItem
              activated={false}
              graphic={
                            Object {
                                          "icon": "x",
                                          "size": "small",
                                        }
              }
              secondaryText="Board 0/1 - Field 0/0"
              text="Move 3"
/>,
            "name": "move-3",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activated": false,
              "graphic": Object {
                "icon": "x",
                "size": "small",
              },
              "secondaryText": "Board 0/1 - Field 0/0",
              "text": "Move 3",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "children": <SimpleListItem
              activated={false}
              graphic={
                            Object {
                                          "icon": "o",
                                          "size": "small",
                                        }
              }
              secondaryText="Board 0/0 - Field 0/1"
              text="Move 2"
/>,
            "name": "move-2",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activated": false,
              "graphic": Object {
                "icon": "o",
                "size": "small",
              },
              "secondaryText": "Board 0/0 - Field 0/1",
              "text": "Move 2",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "children": <SimpleListItem
              activated={false}
              graphic={
                            Object {
                                          "icon": "x",
                                          "size": "small",
                                        }
              }
              secondaryText="Board 0/0 - Field 0/0"
              text="Move 1"
/>,
            "name": "move-1",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activated": false,
              "graphic": Object {
                "icon": "x",
                "size": "small",
              },
              "secondaryText": "Board 0/0 - Field 0/0",
              "text": "Move 1",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
